### PR TITLE
[tests] Attempt at fixing System.ServiceModel.Dispatcher.Bug32886.Bug32886_Test flakiness

### PR DESCRIPTION
**Pull request by ludovic-henry**

### CODE REVIEW

**Comment by akoeplinger** [comment]
We have the same pattern in Bug652331Test.cs and Bug652331_2Test.cs, please fix it there as well.

File: mcs/class/System.ServiceModel/Test/System.ServiceModel.Dispatcher/Bug32886Test.cs [hunk 1]
```diff
@@ -63,16 +63,19 @@ public class Bug32886
...
+
 				client.CelsiusToFahrenheitCompleted += delegate (object o, CelsiusToFahrenheitCompletedEventArgs e) {
-					if (e.Error != null)
-						throw e.Error;
-					Assert.AreEqual ("76.1", e.Result, "#1");
-					wait.Set ();
+					try {
+						if (e.Error != null)
+							throw e.Error;
+						Assert.AreEqual ("76.1", e.Result, "#1");
```
Comment (akoeplinger) [inline mcs/class/System.ServiceModel/Test/System.ServiceModel.Dispatcher/Bug32886Test.cs]: I think we should change our general approach, cause any exception thrown here will not be catched anywhere and just result in a "uncaught exceptions" in the NUnit output (e.g. try changing the assert condition here and you'll see it).

In other tests, I've seen the following pattern:

```csharp
Exception ex = null;
string result = null;
client.CelsiusToFahrenheitCompleted += delegate (object o, CelsiusToFahrenheitCompletedEventArgs e) {

    ex = e.Error;
    result = e.Result;
    wait.Set ();
}
wait.WaitOne ();
Assert.IsNull (ex);
Assert.AreEqual ("76.1", result, "#1");
```

This is much better IMHO.

File: mcs/class/System.ServiceModel/Test/System.ServiceModel.Dispatcher/Bug32886Test.cs [hunk 2]
```diff
@@ -63,16 +63,21 @@ public class Bug32886
...
+					result = e.Result;
 					wait.Set ();
 				};
 
 				client.CelsiusToFahrenheitAsync ("24.5");
-				if (!wait.WaitOne (TimeSpan.FromSeconds (20)))
-					Assert.Fail ("timeout");
+
+				Assert.IsTrue (wait.WaitOne (TimeSpan.FromSeconds (20)), "timeout");
+				Assert.AreEqual (null, error, "#1, inner exception: {0}", error.ToString ());
```
Comment (akoeplinger) [inline mcs/class/System.ServiceModel/Test/System.ServiceModel.Dispatcher/Bug32886Test.cs]: We don't need `error.ToString()` here and in the other tests, NUnit already provides that in the default assert message.

File: mcs/class/System.ServiceModel/Test/System.ServiceModel.Dispatcher/Bug652331_2Test.cs [hunk 3]
```diff
@@ -76,58 +76,73 @@ void RunClient ()
...
+
+				Assert.IsTrue (wait.WaitOne (TimeSpan.FromSeconds (20)), "#1 timeout");
+				Assert.AreEqual (null, error, "#1.1, inner exception: {0}", error.ToString ());
+				Assert.AreEqual ("A", ((DataType1) result).Id, "#1.2");
+			}
+
+			{
+				ManualResetEvent wait = new ManualResetEvent (false);
+				Exception error = null;
+				System.Collections.ObjectModel.ObservableCollection<object> result = null;
```
Comment (akoeplinger) [inline mcs/class/System.ServiceModel/Test/System.ServiceModel.Dispatcher/Bug652331_2Test.cs]: ~System.Collections.ObjectModel.~ObservableCollection to fix the build error.

**Comment by akoeplinger** [comment]
I ran the tests for 3 hours on my machine and they didn't fail, so let's hope this fixes some of the flakiness on Jenkins :)

**Comment by ludovic-henry** [comment]
@monojenkins: merge

**Pull request merged**